src/environment.rs

```diff
@@ -0,0 +1,364 @@
+    /// and if it wasn't, it sets the `current` has the new head of `previous`.
+    /// Then a clone is an empty `current` and the clone of `self.previous`.
+    fn clone(&self) -> Self {
+        if !self.current.is_empty() && !self.was_cloned() {
+            self.previous.replace_with(|old| {
+                Some(Rc::new(Environment {
+                    current: self.current.clone(),
+                    previous: RefCell::new(old.clone()),
+                }))
+            });
+        }
+        Self {
+            current: Rc::new(HashMap::new()),
+            previous: self.previous.clone(),
+        }
+    }
+}
+
+impl<K: Hash + Eq, V: PartialEq> Environment<K, V> {
+    /// Creates a new empty Environment.
+    pub fn new() -> Self {
+        Self {
+            current: Rc::new(HashMap::new()),
+            previous: RefCell::new(None),
+        }
+    }
+
+    /// Inserts a key-value pair into the Environment.
+    /// If the key was present in the _current_ Environment, the previous value is returned.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        if self.was_cloned() {
+            self.current = Rc::new(HashMap::new());
+        }
+        Rc::get_mut(&mut self.current).unwrap().insert(key, value)
+    }
+
+    /// Tries to find the value of a key in the Environment.
+    pub fn get(&self, key: &K) -> Option<V>
+    where
+        V: Clone,
+    {
+        self.iter_layers().find_map(|hmap| hmap.get(key).cloned())
+    }
+
+    /// Creates an iterator that visits all layers from the most recent one to the oldest.
+    /// The element iterator type is `Rc<HashMap<K, V>>`.
+    pub fn iter_layers(&self) -> EnvLayerIter<'_, K, V> {
+        EnvLayerIter {
+            env: if !self.was_cloned() {
+                Some(NonNull::from(self))
+            } else {
+                // if was cloned, current is the same as first of previous (that cannot be empty then)
+                self.previous
+                    .borrow()
+                    .as_ref()
+                    // SAFETY: created from Rc, so cannot be null
+                    .map(|prev| unsafe { NonNull::new_unchecked(Rc::as_ptr(prev) as *mut _) })
+            },
+            _marker: PhantomData,
+        }
+    }
+
+    /// Creates an iterator that visits all elements from the Environment, from the oldest layer
+    /// to the most recent one. It uses this order, so calling `collect` on this iterator to create
+    /// a hashmap would have the same values as the Environment.
+    /// The element iterator type is `(&'env K, &'env V)`, with `'env` being the lifetime of the Environment.
+    pub fn iter_elems(&self) -> EnvElemIter<'_, K, V> {
+        let mut env: Vec<NonNull<HashMap<K, V>>> = self
+            .iter_layers()
+            // SAFETY: all NonNull::new_unchecked comes from pointers created from Rc, so cannot be null
+            .map(|hmap| unsafe { NonNull::new_unchecked(Rc::as_ptr(&hmap) as *mut _) })
+            .collect();
```

Comment:
Why do we need to build a vector here? Can't we just pass directly the mapped `EnvLayerIter` to the element iterator, so that it can fetch the next layer only when needed? Although this is not very important.

Reply:
Nope, we can't, since EnvLayerIter is not a DoubleEndedIterator. Here, during the iteration process, we access the vec frow end to start.